File: crates/traderview-core/src/depreciation_recapture.rs
//! Depreciation recapture on a rental sale — unrecaptured § 1250 gain.
//!
//! Depreciation lowers a rental's basis while you hold it (sheltering rental
//! income); when you sell, that benefit is partly clawed back. For real
//! property depreciated straight-line under MACRS, the depreciation portion
//! of the gain is **unrecaptured § 1250 gain**, taxed at a maximum 25% rate
//! (IRC § 1(h)(1)(E)); the appreciation above the original basis is regular
//! long-term capital gain (0/15/20%).
//!
//!   * adjusted basis = purchase + improvements − accumulated depreciation
//!   * total gain     = (sale price − selling costs) − adjusted basis
//!   * unrecaptured § 1250 = min(accumulated depreciation, total gain)
//!   * LTCG gain      = total gain − unrecaptured § 1250
//!
//! Recapture is limited to the lesser of the depreciation taken or the total
//! gain — a sale at a loss recaptures nothing. NIIT (3.8%) may apply on top
//! but is out of scope here. Pure compute.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
pub struct RecaptureInput {
    pub purchase_price_usd: f64,
    #[serde(default)]
    pub improvements_usd: f64,
    /// Total depreciation taken (or allowable) over the holding period.
    pub accumulated_depreciation_usd: f64,
    pub sale_price_usd: f64,
    #[serde(default)]
    pub selling_costs_usd: f64,
    /// Long-term cap-gains rate on the appreciation portion (e.g. 15 or 20).
    pub ltcg_rate_pct: f64,
    /// Max rate on unrecaptured § 1250 gain (25% cap; lower it if your
    /// ordinary rate is below 25%).
    pub recapture_rate_pct: f64,
}

#[derive(Debug, Clone, Serialize)]
pub struct RecaptureResult {
    pub adjusted_basis_usd: f64,
    pub amount_realized_usd: f64,
    pub total_gain_usd: f64,
    /// Depreciation portion of the gain (taxed at the recapture rate).
    pub unrecaptured_1250_gain_usd: f64,
    /// Appreciation portion above original basis (taxed at LTCG).
    pub ltcg_gain_usd: f64,
    pub recapture_tax_usd: f64,
    pub ltcg_tax_usd: f64,
    pub total_tax_usd: f64,
    /// Total tax as a percent of the total gain.
    pub effective_rate_pct: f64,
    pub after_tax_gain_usd: f64,
    /// True when the sale is at a loss (no gain, no recapture).
    pub is_loss: bool,
}

pub fn analyze(i: &RecaptureInput) -> RecaptureResult {
    let depreciation = i.accumulated_depreciation_usd.max(0.0);
    let adjusted_basis = i.purchase_price_usd + i.improvements_usd - depreciation;
    let amount_realized = i.sale_price_usd - i.selling_costs_usd;
    let total_gain = amount_realized - adjusted_basis;

    if total_gain <= 0.0 {
        return RecaptureResult {
            adjusted_basis_usd: adjusted_basis,
            amount_realized_usd: amount_realized,
            total_gain_usd: total_gain,
            unrecaptured_1250_gain_usd: 0.0,
            ltcg_gain_usd: 0.0,
            recapture_tax_usd: 0.0,
            ltcg_tax_usd: 0.0,
            total_tax_usd: 0.0,
            effective_rate_pct: 0.0,
            after_tax_gain_usd: total_gain,
            is_loss: true,
        };
    }

    // Recapture is the lesser of depreciation taken or the total gain.
    let unrecaptured = depreciation.min(total_gain);
    let ltcg_gain = total_gain - unrecaptured;

    let recapture_tax = unrecaptured * i.recapture_rate_pct / 100.0;
    let ltcg_tax = ltcg_gain * i.ltcg_rate_pct / 100.0;
    let total_tax = recapture_tax + ltcg_tax;
    let effective_rate = if total_gain > 0.0 { total_tax / total_gain * 100.0 } else { 0.0 };

    RecaptureResult {
        adjusted_basis_usd: adjusted_basis,
        amount_realized_usd: amount_realized,
        total_gain_usd: total_gain,
        unrecaptured_1250_gain_usd: unrecaptured,
        ltcg_gain_usd: ltcg_gain,
        recapture_tax_usd: recapture_tax,
        ltcg_tax_usd: ltcg_tax,
        total_tax_usd: total_tax,
        effective_rate_pct: effective_rate,
        after_tax_gain_usd: total_gain - total_tax,
        is_loss: false,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn base() -> RecaptureInput {
        RecaptureInput {
            purchase_price_usd: 200_000.0,
            improvements_usd: 0.0,
            accumulated_depreciation_usd: 50_000.0,
            sale_price_usd: 300_000.0,
            selling_costs_usd: 0.0,
            ltcg_rate_pct: 15.0,
            recapture_rate_pct: 25.0,
        }
    }

    #[test]
    fn basis_gain_and_split() {
        // basis 200k − 50k dep = 150k; gain = 300k − 150k = 150k.
        // unrecaptured = min(50k, 150k) = 50k; LTCG gain = 100k.
        let r = analyze(&base());
        assert!((r.adjusted_basis_usd - 150_000.0).abs() < 1e-6);
        assert!((r.total_gain_usd - 150_000.0).abs() < 1e-6);
        assert!((r.unrecaptured_1250_gain_usd - 50_000.0).abs() < 1e-6);
        assert!((r.ltcg_gain_usd - 100_000.0).abs() < 1e-6);
    }

    #[test]
    fn taxes_split_recapture_and_ltcg() {
        // recapture 50k × 25% = 12.5k; LTCG 100k × 15% = 15k; total 27.5k.
        let r = analyze(&base());
        assert!((r.recapture_tax_usd - 12_500.0).abs() < 1e-6);
        assert!((r.ltcg_tax_usd - 15_000.0).abs() < 1e-6);
        assert!((r.total_tax_usd - 27_500.0).abs() < 1e-6);
    }

    #[test]
    fn recapture_limited_to_total_gain() {
        // Depreciation 50k but total gain only 30k → recapture caps at 30k, LTCG 0.
        let r = analyze(&RecaptureInput { sale_price_usd: 180_000.0, ..base() });
        // basis 150k; gain = 180k − 150k = 30k.
        assert!((r.total_gain_usd - 30_000.0).abs() < 1e-6);
        assert!((r.unrecaptured_1250_gain_usd - 30_000.0).abs() < 1e-6);
        assert!(r.ltcg_gain_usd.abs() < 1e-6);
    }

    #[test]
    fn sale_at_loss_recaptures_nothing() {
        // Sell for 140k < adjusted basis 150k → loss.
        let r = analyze(&RecaptureInput { sale_price_usd: 140_000.0, ..base() });
        assert!(r.is_loss);
        assert!(r.total_tax_usd.abs() < 1e-9);
        assert!(r.total_gain_usd < 0.0);
    }

    #[test]
    fn selling_costs_reduce_amount_realized() {
        let r = analyze(&RecaptureInput { selling_costs_usd: 18_000.0, ..base() });
        assert!((r.amount_realized_usd - 282_000.0).abs() < 1e-6);
        // gain = 282k − 150k = 132k.
        assert!((r.total_gain_usd - 132_000.0).abs() < 1e-6);
    }

    #[test]
    fn improvements_raise_basis() {
        let r = analyze(&RecaptureInput { improvements_usd: 40_000.0, ..base() });
        // basis = 200k + 40k − 50k = 190k; gain = 300k − 190k = 110k.
        assert!((r.adjusted_basis_usd - 190_000.0).abs() < 1e-6);
        assert!((r.total_gain_usd - 110_000.0).abs() < 1e-6);
    }

    #[test]
    fn zero_depreciation_is_all_ltcg() {
        let r = analyze(&RecaptureInput { accumulated_depreciation_usd: 0.0, ..base() });
        assert!(r.unrecaptured_1250_gain_usd.abs() < 1e-9);
        assert!((r.ltcg_gain_usd - r.total_gain_usd).abs() < 1e-9);
    }

    #[test]
    fn effective_rate_between_ltcg_and_recapture() {
        let r = analyze(&base());
        // 27.5k / 150k = 18.33% — between 15% (LTCG) and 25% (recapture).
        assert!((r.effective_rate_pct - (27_500.0 / 150_000.0 * 100.0)).abs() < 1e-9);
        assert!(r.effective_rate_pct > 15.0 && r.effective_rate_pct < 25.0);
    }
}

File: frontend/js/views/depreciation_recapture.js
// Depreciation recapture — splits a rental sale's gain into unrecaptured
// § 1250 gain (max 25%) and LTCG, with the tax on each, via
// /calc/depreciation-recapture. Updates live as you type.

import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';
import * as enh from '../calc_enhance.js';

const VIEW = 'depreciation-recapture';
let lastReport = null;
let lastBody = null;

const FIELDS = [
    ['purchase_price_usd', 'Purchase price ($)', 200000],
    ['improvements_usd', 'Improvements ($)', 0],
    ['accumulated_depreciation_usd', 'Accumulated depreciation ($)', 50000],
    ['sale_price_usd', 'Sale price ($)', 300000],
    ['selling_costs_usd', 'Selling costs ($)', 0],
    ['ltcg_rate_pct', 'LTCG rate (%)', 15],
    ['recapture_rate_pct', 'Recapture rate (max 25%)', 25],
];

const money = (n) => '$' + Number(n).toLocaleString(undefined, { maximumFractionDigits: 0 });
const pct = (n) => Number(n).toFixed(2) + '%';

export async function renderDepreciationRecapture(mount, _appState) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.dr.h1.title">// DEPRECIATION RECAPTURE</span></h1>
        <p class="muted small" data-i18n="view.dr.hint.intro">
            Depreciation lowers a rental's basis while you hold it; selling claws part of that
            back. For real property, the depreciation portion of the gain is unrecaptured
            § 1250 gain, taxed at a maximum 25% rate; the appreciation above your original
            basis is regular long-term capital gain. Recapture is limited to the lesser of
            depreciation taken or the total gain — a sale at a loss recaptures nothing. NIIT
            (3.8%) may apply on top. Updates as you type.
        </p>
        <div class="lpv-split">
        <div class="chart-panel">
            <h2 data-i18n="view.dr.h2.inputs">The sale</h2>
            <form id="dr-form" class="inline-form">
                ${FIELDS.map(([key, label, def]) => `
                    <label><span data-i18n="view.dr.label.${key}">${label}</span>
                        <input type="number" step="0.01" min="0" name="${key}" value="${def}" required></label>
                `).join('')}
            </form>
            <div id="dr-tools" class="ce-toolbar"></div>
        </div>
        <div id="dr-result" class="lpv-preview"></div>
        </div>
    `;
    applyUiI18n(mount);

    const form = mount.querySelector('#dr-form');
    enh.prefillForm(form, enh.readHashInputs());
    const readBody = () => {
        const fd = new FormData(form);
        const body = {};
        for (const [key] of FIELDS) body[key] = Number(fd.get(key)) || 0;
        return body;
    };
    const generate = async () => {
        const body = readBody();
        try {
            const r = await api.calcDepreciationRecapture(body);
            if (!viewIsCurrent(tok)) return;
            lastReport = r; lastBody = body;
            renderResult(mount, r);
        } catch (err) {
            showToast(err.message || t('view.dr.toast.error'), { level: 'error' });
        }
    };
    enh.mountToolbar(mount.querySelector('#dr-tools'), { viewId: VIEW, getInputs: () => lastBody || readBody(), getRows: () => reportRows(lastReport), filename: 'depreciation-recapture.csv' });
    form.addEventListener('input', debounce(generate, 250));
    form.addEventListener('submit', (e) => { e.preventDefault(); generate(); });
    generate();
}

function reportRows(r) {
    if (!r) return [];
    return [
        ['metric', 'value'],
        ['total_gain_usd', r.total_gain_usd],
        ['unrecaptured_1250_gain_usd', r.unrecaptured_1250_gain_usd],
        ['recapture_tax_usd', r.recapture_tax_usd],
        ['ltcg_gain_usd', r.ltcg_gain_usd],
        ['ltcg_tax_usd', r.ltcg_tax_usd],
        ['total_tax_usd', r.total_tax_usd],
    ];
}

function renderResult(mount, r) {
    const el = mount.querySelector('#dr-result');
    if (r.is_loss) {
        el.innerHTML = `<div class="chart-panel"><h2 data-i18n="view.dr.h2.result">The tax</h2>
            <div class="cards"><div class="card"><div class="label" data-i18n="view.dr.card.gain">Total gain</div>
                <div class="value neg">${money(r.total_gain_usd)}</div></div></div>
            <p class="muted small" data-i18n="view.dr.loss">Sale is at a loss — no gain and no depreciation recapture.</p></div>`;
        applyUiI18n(el);
        return;
    }
    el.innerHTML = `
        <div class="chart-panel">
            <h2 data-i18n="view.dr.h2.result">The tax</h2>
            <div class="cards">
                <div class="card neg"><div class="label" data-i18n="view.dr.card.total_tax">Total tax</div>
                    <div class="value neg">${money(r.total_tax_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.dr.card.effective">Effective rate</div>
                    <div class="value">${pct(r.effective_rate_pct)}</div></div>
                <div class="card pos"><div class="label" data-i18n="view.dr.card.after_tax">After-tax gain</div>
                    <div class="value pos">${money(r.after_tax_gain_usd)}</div></div>
            </div>
            ${enh.svgBarChart([
                { label: '§1250 tax', value: -r.recapture_tax_usd },
                { label: 'LTCG tax', value: -r.ltcg_tax_usd },
                { label: 'After-tax', value: r.after_tax_gain_usd },
            ])}
            <table class="data-table">
                <thead><tr>
                    <th data-i18n="view.dr.col.line">Line</th>
                    <th data-i18n="view.dr.col.gain">Gain</th>
                    <th data-i18n="view.dr.col.tax">Tax</th>
                </tr></thead>
                <tbody>
                    <tr><td data-i18n="view.dr.row.recapture">Unrecaptured § 1250 (≤25%)</td>
                        <td>${money(r.unrecaptured_1250_gain_usd)}</td><td>${money(r.recapture_tax_usd)}</td></tr>
                    <tr><td data-i18n="view.dr.row.ltcg">LTCG (appreciation)</td>
                        <td>${money(r.ltcg_gain_usd)}</td><td>${money(r.ltcg_tax_usd)}</td></tr>
                    <tr class="emph"><td data-i18n="view.dr.row.total">Total gain</td>
                        <td>${money(r.total_gain_usd)}</td><td>${money(r.total_tax_usd)}</td></tr>
                    <tr><td data-i18n="view.dr.row.basis">Adjusted basis</td>
                        <td colspan="2">${money(r.adjusted_basis_usd)}</td></tr>
                </tbody>
            </table>
        </div>
    `;
    applyUiI18n(el);
}
